pkg/promtail/targets/filetarget.go

```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
+	toStartWatching := missing(t.watches, dirs)
+	t.startWatching(toStartWatching)
 
-		path:     path,
-		filename: filename,
-		tail:     tail,
-		quit:     make(chan struct{}),
-		done:     make(chan struct{}),
-	}
-	go tailer.run()
-	filesActive.Add(1.)
-	return tailer, nil
-}
+	// Remove any directories which no longer need watching.
+	toStopWatching := missing(dirs, t.watches)
+	t.stopWatching(toStopWatching)
 
-func (t *tailer) run() {
-	level.Info(t.logger).Log("msg", "start tailing file", "path", t.path)
-	positionSyncPeriod := t.positions.SyncPeriod()
-	positionWait := time.NewTicker(positionSyncPeriod)
+	// fsnotify.Watcher doesn't allow us to see what is currently being watched so we have to track it ourselves.
+	t.watches = dirs
 
-	defer func() {
-		level.Info(t.logger).Log("msg", "stopped tailing file", "path", t.path)
-		positionWait.Stop()
-		err := t.markPosition()
-		if err != nil {
-			level.Error(t.logger).Log("msg", "error getting tail position", "path", t.path, "error", err)
-		}
-		close(t.done)
-	}()
+	t.updateTailers(matches)
 
-	for {
-		select {
-		case <-positionWait.C:
-			err := t.markPosition()
-			if err != nil {
-				level.Error(t.logger).Log("msg", "error getting tail position", "path", t.path, "error", err)
-				continue
-			}
+	return nil
+}
 
-		case line, ok := <-t.tail.Lines:
-			if !ok {
-				return
-			}
+func (t *FileTarget) startWatching(dirs map[string]struct{}) {
+	for dir := range dirs {
+		if _, ok := t.watches[dir]; ok {
+			continue
+		}
+		level.Debug(t.logger).Log("msg", "watching new directory", "directory", dir)
+		if err := t.watcher.Add(dir); err != nil {
+			level.Error(t.logger).Log("msg", "error adding directory to watcher", "error", err)
+		}
+	}
+}
 
-			if line.Err != nil {
-				level.Error(t.logger).Log("msg", "error reading line", "path", t.path, "error", line.Err)
+func (t *FileTarget) stopWatching(dirs map[string]struct{}) {
+	for dir := range dirs {
+		if _, ok := t.watches[dir]; !ok {
+			continue
+		}
+		level.Debug(t.logger).Log("msg", "removing directory from watcher", "directory", dir)
+		err := t.watcher.Remove(dir)
+		if err != nil {
+			level.Error(t.logger).Log("msg", " failed to remove directory from watcher", "error", err)
+		}
+		// Shutdown and cleanup and tailers for files in directories no longer being watched.
+		for tailedFile := range t.tails {
+			if filepath.Dir(tailedFile) == dir {
+				t.stopTailing(tailedFile)
 			}
+		}
+	}
+}
 
-			readLines.WithLabelValues(t.path).Inc()
-			readBytes.WithLabelValues(t.path).Add(float64(len(line.Text)))
-			if err := t.handler.Handle(model.LabelSet{}, line.Time, line.Text); err != nil {
-				level.Error(t.logger).Log("msg", "error handling line", "path", t.path, "error", err)
-			}
-		case <-t.quit:
-			return
+func (t *FileTarget) updateTailers(matches []string) {
+	// Start tailing all of the matched files if not already doing so.
+	for _, p := range matches {
+		err := t.startTailing(p)
```

Comment:
Is it safe to start tailing things we already tail?  Oh I see it is.  Can you use the same pattern as in `sync()`?

Reply:
can you elaborate a little with what you mean about using the same pattern as sync()?